### PR TITLE
feat: cuTile Grouped MXFP8 Quantization

### PR DESCRIPTION
**Pull request by philipphack**
## 📌 Description

Adds `mxfp8_grouped_quantize`, a masked grouped MXFP8 quantizer for the MoE activation path. It quantizes a `[B, M, K]` batch to FP8 E4M3 with UE8M0 block scales and emits the standard 128×4 swizzled scale-factor layout consumed by a Blackwell masked grouped GEMM.

The cuTile kernel lives under `flashinfer/quantization/kernels/cutile/`, matching the existing cuTile GEMM backends (#3426). Availability detection is hoisted into a shared `flashinfer/cutile/` package (`is_cuda_tile_available()`), and the CuTe-DSL quantization kernels in `quantization/kernels/__init__.py` are now imported lazily so this path does not pull in `nvidia-cutlass-dsl`.

Also adds an `mxfp8_grouped_quantize` `TraceTemplate` + reference-correctness test, public exports, and a `docs/api/quantization.rst` entry. Requires SM100+ (Blackwell).

## 🔍 Related Issues

N/A

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Local (all passed): `pre-commit run --all-files`; `pytest tests/utils/test_fp8_quantize.py -k grouped`; `pytest tests/trace/ -k grouped` (SM100 + cuda.tile).

New coverage: exact per-group values + unswizzled scales; 128×4 GEMM layout/stride contract; empty experts (`mask=0`); CUDA-graph capture/replay; fake/meta op; dequantized round-trip reference test.

## Reviewer Notes

- The trace template intentionally has no `reference`: a plain-layout reference is not comparable to the swizzled/permuted output, so correctness is verified by a dequantized round-trip test rather than an embedded oracle.

## Summary by CodeRabbit

* **New Features**
  * Added grouped MXFP8 quantization via `mxfp8_grouped_quantize` exposed at the package top level (SM100+ required; `K` must be divisible by 32).

* **Documentation**
  * Extended quantization API docs with the new grouped quantization function, backend notes, and input/shape constraints.

* **Tests**
  * Added grouped MXFP8 quantization coverage, including reference correctness, trace generation/fixtures, CUDA graph capture, empty-group handling, concurrency, and output scale/layout validation.

### CODE REVIEW

File: flashinfer/quantization/kernels/cutile/mxfp8_grouped_quantize_cutile.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+    num_groups = problem_sizes.size(0)
+    if num_groups > MAX_GROUPS_FUSED:
+        raise ValueError(
+            f"num_groups={num_groups} exceeds MAX_GROUPS_FUSED={MAX_GROUPS_FUSED}."
+        )
+    if num_groups == 0:
+        return _PrefixSchedule(
+            torch.zeros(1, dtype=torch.int32, device=input.device),
+            total_tiles=-1,
+        )
```
[the remark this answers is not in the document]
Comment (philipphack) [reply flashinfer/quantization/kernels/cutile/mxfp8_grouped_quantize_cutile.py:102]: `num_groups == 0` only occurs in the degenerate case of zero experts, which has no practical performance implications.

File: flashinfer/quantization/kernels/cutile/mxfp8_grouped_quantize_cutile.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+        raise ValueError("blocks_per_sm must be positive")
+
+    _check_common_args(
+        input,
+        problem_sizes,
+        expert_offsets,
+        blockscale_offsets,
+        quant_output,
+        scale_factor,
+    )
```
[the remark this answers is not in the document]
Comment (philipphack) [reply flashinfer/quantization/kernels/cutile/mxfp8_grouped_quantize_cutile.py:188]: `mxfp8_grouped_quantize_cutile_with_prefix_schedule` is a public entry point that can be called directly, so it must validate its own inputs; the check is host-side only, runs once per launch (never during graph replay), and is negligible next to the kernel launch, so this is not a meaningful hot-path cost.

File: flashinfer/quantization/kernels/cutile/mxfp8_grouped_quantize_cutile.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+            torch.zeros(1, dtype=torch.int32, device=input.device),
+            total_tiles=-1,
+        )
+
+    if input.device.index is None:
+        device_index = torch.cuda.current_device()
+    else:
+        device_index = input.device.index
+    cache_key = (input.device.type, device_index)
+    tile_offsets = _FUSED_PREFIX_GRAPH_TILE_OFFSETS.get(cache_key)
```
Comment (kahyunnam) [inline flashinfer/quantization/kernels/cutile/mxfp8_grouped_quantize_cutile.py]: This might become a concurrency issue for multiple streams on the same device, since tile_offsets is indexed by only device type/index. It could lead to a similar bug as https://github.com/flashinfer-ai/flashinfer/issues/3618. 

A fix could be to also index this by per-stream: mirroring _get_cache_buf_ring from [#3625 ](https://github.com/flashinfer-ai/flashinfer/pull/3625)

**Review by kahyunnam — COMMENTED** [review]
One concern below, but LGTM otherwise!

File: tests/utils/test_fp8_quantize.py [hunk 4]
```diff
@@ -55,6 +83,281 @@ def test_mxfp8_quantize_torch(m, k, dtype, is_sf_swizzled_layout, device, backen
...
 
+@pytest.mark.parametrize("batch_shape", [(1, 120, 64), (2, 128, 128), (3, 256, 160)])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
+@torch.inference_mode()
+def test_mxfp8_grouped_quantize(batch_shape, dtype):
+    if not torch.cuda.is_available():
+        pytest.skip("CUDA is not available")
+    major, _ = get_compute_capability(torch.device("cuda:0"))
+    if major < 10:
+        pytest.skip("mxfp8 grouped quantization requires compute capability >= 10")
```
[the remark this answers is not in the document]
Comment (kahyunnam) [reply tests/utils/test_fp8_quantize.py]: @philipphack this seems like a stylistic nit, but is probably a nice to have, could you take a look?

**Review by kahyunnam — APPROVED** [review]
LGTM, thanks for the contribution!

**Review by bkryu — APPROVED** [review]
